File: 11-Container With Most Water.py
'''
class Solution:
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        max_space = 0
        for start in range(len(height)):
            for end in range(start + 1, len(height)):
                max_space = max(max_space, min(height[start], height[end]) * (end - start))
        return max_space
'''


class Solution:
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        max_space, left, right = 0, 0, len(height) - 1
        while left < right:
            max_space = max(max_space, min(height[left], height[right]) * (right - left))
            if height[left] < height[right]: 
                left += 1 
            else:
                right -= 1
        return max_space

File: C++/11-Container With Most Water.cpp
// Runtime: 20 ms, faster than 71.13% of C++ online submissions for Container With Most Water.
// Memory Usage: 9.7 MB, less than 87.35% of C++ online submissions for Container With Most Water.
class Solution {
public:
    int maxArea(vector<int>& height) {
        int max_space = 0, l = 0, r = height.size() - 1;
        while (l < r) {
            max_space = max(max_space, min(height[l], height[r]) * (r - l));
            if (height[l] > height[r])
                --r;
            else
                ++l;
        }
        return max_space;
    }
};